File: 347-TopKFrequentElements/347-TopKFrequentElements.java
// Last updated: 18/08/2026, 14:56:21
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        List<Integer>[] bucket = new List[nums.length + 1];
        Arrays.setAll(bucket, i -> new ArrayList<>());
        HashMap<Integer, Integer> freq = new HashMap<>();

        for (int num : nums) {
            int value = freq.getOrDefault(num, 0);
            freq.put(num, value + 1);
        }

        freq.forEach((key, value) -> {
            bucket[value].add(key);
        });

        int[] res = new int[k];
        int index = 0;

        for (int i = bucket.length - 1; i > 0; i--) {
            for (int j : bucket[i]) {
                res[index] = j;
                index++;

                if (index >= k) {
                    return res;
                }
            }
        }

        return null;
    }
}

File: 347-TopKFrequentElements/347-TopKFrequentElements.py
# Last updated: 18/08/2026, 14:56:25
class Solution:
    def topKFrequent(self, nums: List[int], k: int) -> List[int]:
        freq = {}
        bucket = [[] for x in range(len(nums) + 1)]

        for num in nums:
            freq[num] = 1 + freq.get(num, 0)
        
        for key, value in freq.items():
            bucket[value].append(key)
        
        res = []
        for i in range(len(bucket) - 1, 0, -1):
            for j in range(len(bucket[i])):
                res.append(bucket[i][j])

                if len(res) == k: return res
